{% extends "Messages/modal_template.html" %}
{% block id %}About{% endblock %}
{% block title %}
    About mirnaQC
{% endblock %}
{% block text %}
<p> miRNA-seq quality control is frequently restricted to calculating
    PhredScores and filtering of low complexity reads since miRNA specific quality measures cannot be easily assessed.</p>
    <p> mirnaQC provides several quality features that can help researchers identify issues in their samples. These features are
    provided as absolute values and ranked with a percentile calculated from a corpus of more than 36,000 samples.

    </p>
    <p>Please note that mirnaQC was developed and tested using Google Chrome and Mozilla Firefox. Using other (older) browsers
    will likely result in some features not working properly. Please contact us to try to solve them.</p>
    <p>mirnaQC was developed by the Computational Epigenomics Group at University of Granada</p>
{#    <ul>#}
{#  <li>Detect the adapter sequence: TGGAATTCTCGGGTGCCAAGGG </li>#}
{#  <li>Minimum length of adapter sequence that needs to be aligned: 10 </li>#}
{#  <li>Allowed mismatches between adapter and reference sequences: 1 </li>#}
{#    </ul>#}


{% endblock %}
